src/main/java/de/zalando/aruha/nakadi/service/ClosedConnectionsCrutch.java

```diff
@@ -0,0 +1,294 @@
+            this.connectionInfo = connectionInfo;
+            this.callback = callback;
+        }
+
+        @Override
+        public int compareTo(final ListenerInfo o) {
+            return addedAt.compareTo(o.addedAt);
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            final ListenerInfo that = (ListenerInfo) o;
+
+            if (addedAt != that.addedAt) return false;
+            if (!connectionInfo.equals(that.connectionInfo)) return false;
+            return callback.equals(that.callback);
+
+        }
+
+        @Override
+        public int hashCode() {
+            final long value = addedAt.getTime();
+            return (int) (value ^ (value >>> 32));
+        }
+    }
+
+    private final int port;
+    private final Map<ConnectionInfo, List<BooleanSupplier>> listeners = new HashMap<>();
+    private final SortedSet<ListenerInfo> toAdd = new TreeSet<>();
+    private final Meter meterClosed;
+    private final FeatureToggleService featureToggleService;
+    private static final Logger LOG = LoggerFactory.getLogger(ClosedConnectionsCrutch.class);
+    private static final long DELAY_MILLIS = 100;
+
+    private static final String FEATURE_NAME = "close_crutch";
+    @Autowired
+    public ClosedConnectionsCrutch(
+            @Value("${server.port}") final int port,
+            final MetricRegistry metricRegistry,
+            final FeatureToggleService featureToggleService) {
+        this.port = port;
+        this.meterClosed = metricRegistry.meter("nakadi.close_crutch.closed");
+        this.featureToggleService = featureToggleService;
+    }
+
+    public void listenForConnectionClose(
+            final InetAddress address,
+            final int port,
+            final BooleanSupplier onCloseListener) {
+        if (!featureToggleService.isFeatureEnabled(FEATURE_NAME)) {
+            return;
+        }
+        LOG.debug("Listening for connection to close using crutch (" + address + ":" + port + ")");
+        final Date now = new Date();
+        synchronized (this) {
+            toAdd.add(new ListenerInfo(now, new ConnectionInfo(address, port), onCloseListener));
+        }
+    }
+
+    @Scheduled(fixedDelay = 1000)
+    public void refresh() throws IOException {
+        if (!featureToggleService.isFeatureEnabled(FEATURE_NAME)) {
+            return;
+        }
+        final long selector = System.currentTimeMillis() - DELAY_MILLIS;
+        synchronized (this) {
+            final Iterator<ListenerInfo> it = toAdd.iterator();
+            while (it.hasNext()) {
+                final ListenerInfo first = it.next();
+                if (first.addedAt.getTime() >= selector) {
+                    break;
+                }
+                listeners.computeIfAbsent(first.connectionInfo, info -> new ArrayList<>()).add(first.callback);
```

Comment:
There is no way it can grow. 
Each time on refresh() call there is a for cycle, that goes thru all the connections in listeners collection. 
If connection must be removed, than it is removed with notification. If it's not - than connection is still open.

Maybe you misunderstood the way it is removed:
```
listeners.remove(key)
```